### PR TITLE
Make keyboard strategies resilient to repeated invocation.

### PR DESCRIPTION
**Pull request by seanparsons**
**Problem:**
Currently the keyboard strategies are implemented in an instantaneous style, such that they handle a keypress at a point in time. This can cause issues however when the strategies are invoked more than once when for example a modifier is pressed or released as it may cause the logic to internally move something a second time and then remove a guideline that was being shown.

**Fix:**
The primary change here is to record all the changes that happen to the keypresses since the interaction started. Then both the absolute move and resize strategies have a pass that accumulates repeated presses with counts. That is then followed by a strategy specific approach to combine those accumulated keypresses into a minimum number of commands for the strategy.

**Commit Details:**
- `KeyboardInteractionData` now records an array of `KeyState` elements,
  which include all the keypresses triggered since the interaction was first
  started.
- `interactionDataHardReset` now prunes the keyboard interaction key states
  to the latest.
- Extracted out `getDragDeltaFromKey` function from the two keyboard strategies.
- Added shared utility functions `getLastKeyPressState` and `accumulatePresses`.
- The move and resize strategies now accumulate repeated presses, populating a
  count of how many repeated presses for identical keypresses there are. Then
  the resultant drag is accumulated from those changes, so that the minimum of
  commands is produced by the strategy.
- Modified a few conditionals to correctly handle the case that now there is
  not a common `modifiers` field on both keyboard and mouse interactions.
- `handleAccumulatingKeypresses` now also culls the key states so that they
  are not carried over into a different strategy.
- Added some `keyup` dispatches to some test code, because we were firing repeated `keydown`
  events which to our logic looks like they are still depressed when we press
  another key.

### CODE REVIEW

File: editor/src/components/canvas/canvas-strategies/interaction-state.ts [hunk 1]
```diff
@@ -340,6 +338,32 @@ export function hasDragModifiersChanged(
   )
 }
 
+export function hasKeyboardModifierChanged(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/interaction-state.ts:341]: it looks like this function is not used anywhere, only imported in dispatch-strategies.tsx

File: editor/src/components/canvas/canvas-strategies/keyboard-absolute-move-resize.spec.browser2.tsx [hunk 2]
```diff
@@ -44,6 +44,34 @@ describe('Keyboard Absolute Strategies E2E', () => {
     await expectElementLeftInPrintedCode(30)
   })
 
+  it('Pressing Shift + ArrowRight 3 times, then pressing ArrowRight 3 times', async () => {
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/keyboard-absolute-move-resize.spec.browser2.tsx:47]: yay more keypress tests! 😎

File: editor/src/components/editor/store/dispatch-strategies.tsx [hunk 3]
```diff
@@ -406,50 +411,68 @@ function handleAccumulatingKeypresses(
...
-      customStrategyState: strategyResult.customState ?? strategyState.customStrategyState,
-      startingAllElementProps: strategyState.startingAllElementProps,
-    }
+    const interactionData = newEditorState.canvas.interactionSession.interactionData
+    if (isKeyboardInteractionData(interactionData)) {
+      const lastKeyState = last(interactionData.keyStates)
+      const updatedInteractionData: KeyboardInteractionData = {
+        ...interactionData,
+        keyStates: lastKeyState == null ? [] : [lastKeyState],
+      }
```
Comment (gbalint) [inline editor/src/components/editor/store/dispatch-strategies.tsx:420]: I have ambigous feelings about this. Until now the data flow looked like this:
- user interactions changed the interaction state
- based on the editor state and the interaction state strategies modified the state, but mostly the projectcontents itself and/or the StrategyState

But in this code the strategy handling modifies the interaction state itself, so the interaction state is not purely a description of the user behavior. I believe when the strategies need state, that should be in the StrategyState not in the InteractionState. So how I would do it is to put this information into the StrategyState, or to be more precise, probably into the StrategyState.customStrategyState.

File: editor/src/components/canvas/canvas-strategies/absolute-resize-bounding-box-strategy.tsx [hunk 4]
```diff
@@ -51,6 +51,8 @@ export const absoluteResizeBoundingBoxStrategy: CanvasStrategy = {
     } else {
       return false
     }
+
+    return false
```
[the remark this answers is not in the document]
Comment (seanparsons) [reply editor/src/components/canvas/canvas-strategies/absolute-resize-bounding-box-strategy.tsx]: Sure is, fixed in 44845db84.

File: editor/src/components/editor/store/dispatch-strategies.tsx (hunk 3 above)
Comment (seanparsons) [reply editor/src/components/editor/store/dispatch-strategies.tsx:419]: Surely `interactionDataHardReset` would fall into that as well? As that is triggered from `interactionHardReset` and it modifies the `drag` and `prevDrag` values.

Comment (gbalint) [reply editor/src/components/editor/store/dispatch-strategies.tsx:419]: Good argument! It seems the separation between interaction and strategy states is just not that clear

Comment (balazsbajorics) [reply editor/src/components/editor/store/dispatch-strategies.tsx:419]: I'd definitely not put it in the `customStrategyState`, that should be reserved for things that are set by apply functions.

**Review by balazsbajorics — APPROVED** [review]

**Pull request merged**